### PR TITLE
Laurentnegre/ch6009/rajouter la possibilité de changer de fuseau

### PR DESCRIPTION
**Pull request by Raulnet**

### CODE REVIEW

File: src/Entity/BaseEvent.php [hunk 1]
```diff
@@ -101,6 +101,16 @@
      */
     protected $description;
 
+    /**
+     * @var string|null
+     *
+     * @ORM\Column(length=50)
+     *
+     * @JMS\Groups({"public", "event_read", "citizen_action_read"})
+     * @JMS\SerializedName("timeZone")
```
Comment (ottaviano) [inline src/Entity/BaseEvent.php]: no need it

Comment (Raulnet) [reply src/Entity/BaseEvent.php]: removed

File: src/Admin/EventAdmin.php [hunk 2]
```diff
@@ -65,61 +66,64 @@ protected function configureShowFields(ShowMapper $showMapper)
     {
         $showMapper
             ->with('Événement', ['class' => 'col-md-7'])
-                ->add('name', null, [
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Admin/EventAdmin.php]: edited

File: app/migrations/Version20181228122041.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
```
Comment (JeremieSamson) [inline app/migrations/Version20181228122041.php]: you can remove this line

File: src/Entity/BaseEvent.php [hunk 4]
```diff
@@ -101,6 +101,13 @@
      */
     protected $description;
 
+    /**
+     * @var string|null
+     *
+     * @ORM\Column(length=50)
```
Comment (JeremieSamson) [inline src/Entity/BaseEvent.php]: in your getter `timeZone` could be null, but not here, is there a mistake ?

File: src/DataFixtures/ORM/LoadCitizenActionData.php [hunk 5]
```diff
@@ -59,6 +60,7 @@ public function load(ObjectManager $manager)
             'begin_at' => (new Chronos('+9 days'))->setTime(9, 00, 00, 000),
             'finish_at' => (new Chronos('+9 days'))->setTime(19, 00, 00, 000),
             'capacity' => 30,
+            'time_zone' => 'Europe/Paris',
```
Comment (JeremieSamson) [inline src/DataFixtures/ORM/LoadCitizenActionData.php]: Isn't `'Europe/Paris'` the default timezone ?

File: src/DataFixtures/ORM/LoadEventData.php [hunk 6]
```diff
@@ -94,6 +94,7 @@ public function load(ObjectManager $manager)
             'begin_at' => (new Chronos('+3 days'))->format('Y-m-d').' 09:30:00',
             'finish_at' => (new Chronos('+3 days'))->format('Y-m-d').' 19:00:00',
             'capacity' => 10,
+            'time_zone' => 'Europe/Paris',
```
Comment (JeremieSamson) [inline src/DataFixtures/ORM/LoadEventData.php]: Isn't `'Europe/Paris'` the default timezone ?

File: src/Entity/BaseEvent.php (hunk 4 above)
Comment (Raulnet) [reply src/Entity/BaseEvent.php]: corriger

File: app/migrations/Version20181228122041.php (hunk 3 above)
Comment (Raulnet) [reply app/migrations/Version20181228122041.php]: solved

File: src/DataFixtures/ORM/LoadCitizenActionData.php (hunk 5 above)
Comment (Raulnet) [reply src/DataFixtures/ORM/LoadCitizenActionData.php]: the timeZone he's not required for loadData

File: src/DataFixtures/ORM/LoadEventData.php (hunk 6 above)
Comment (Raulnet) [reply src/DataFixtures/ORM/LoadEventData.php]: the timeZone is not required for build Event

File: src/Entity/BaseEvent.php [hunk 7]
```diff
@@ -101,6 +101,13 @@
      */
     protected $description;
 
+    /**
+     * @var string
+     *
+     * @ORM\Column(length=50)
+     */
+    protected $timeZone;
```
Comment (Remg) [inline src/Entity/BaseEvent.php]: Does this need to be exposed in apis ?
Does this need to be synced with api.en-marche.fr ?

If yes, you should add annotations regarding serialization (JMS\Group & SerializedName, you can check the beginAt & finishAt properties for examples.

Comment (Raulnet) [reply src/Entity/BaseEvent.php]: expose in api Added

File: src/Form/BaseEventCommandType.php [hunk 8]
```diff
@@ -30,6 +31,9 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'purifier_type' => 'enrich_content',
             ])
             ->add('address', AddressType::class)
+            ->add('timeZone', TimezoneType::class, [
+                'choices' => $this->getTimezones(),
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Form/BaseEventCommandType.php]: I need to display the offsett ``` Paris (UTC +1) ```

File: src/Entity/BaseEvent.php [hunk 9]
```diff
@@ -254,34 +261,24 @@ public function setName(string $name): void
...
+    public function setTimeZone(?string $timeZone): void
+    {
+        $this->timeZone = $timeZone;
+    }
+
+    public function isFinished(): bool
+    {
+        $finishAt = new \DateTimeImmutable(
+            $this->finishAt->format('Y-m-d H:i'),
+            $timezone = new \DateTimeZone($this->timeZone)
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Entity/BaseEvent.php]: That's doesn't work. the property was load with the TimeZone of the Server 'Europe/Paris'. I need to compare the date with a same TimeZone. If you compare 2 dateTime with different timeZone the comparator Resolve automaticly the timeZone

File: src/Address/GeoCoder.php [hunk 10]
```diff
@@ -29,4 +29,17 @@ public function getCountryCodeFromIp(string $ip): ?string
...
+    {
+        try {
+            return  $this->provider->city($ip)->location->timeZone ?? 'Europe/Paris';
+        } catch (GeoIp2Exception $e) {
+            $this->logger->warning(sprintf('[GeoIP2] Unable to locate IP [%s]: %s', $ip, $e->getMessage()), [
+                'stack_trace' => $e->getTraceAsString(),
+            ]);
+        }
+
+        return 'Europe/Paris';
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Address/GeoCoder.php]: done, I use the const from GeoCoder.

File: src/Event/EventCommand.php [hunk 11]
```diff
@@ -44,9 +44,10 @@ public function __construct(
         \DateTimeInterface $beginAt = null,
         \DateTimeInterface $finishAt = null,
         bool $isForLegislatives = false,
-        Event $event = null
+        Event $event = null,
+        string $timezone = 'Europe/Paris'
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Event/EventCommand.php]: done

File: src/Event/EventFactory.php [hunk 12]
```diff
@@ -72,8 +79,14 @@ public function createCitizenActionFromArray(array $data): CitizenAction
             $data['description'],
             $data['address'],
             $data['begin_at'],
-            $data['finish_at']
+            $data['finish_at'],
+            0
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Event/EventFactory.php]: old code remove

File: src/Twig/EventExtension.php [hunk 13]
```diff
@@ -13,4 +14,11 @@ public function getFunctions()
             new TwigFunction('is_event_already_participating', [EventRuntime::class, 'isEventAlreadyParticipating']),
         ];
     }
+
+    public function getFilters()
+    {
+        return [
+            new TwigFilter('utcOffset', [EventRuntime::class, 'offsetTimeZone']),
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Twig/EventExtension.php]: done

**Pull request merged**